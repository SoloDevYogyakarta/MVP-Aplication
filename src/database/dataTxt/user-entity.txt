{"id":1088,"public_id":"io7NL8MF7mPHdla16p3X0","username":"updated-Opal_Rempel28","email":"Dameon_Cronin16@gmail.com","phone_number":null,"password":"$2b$10$ix8JcUoVPanbpNiAczvphut0WyjV/wcUUGw0/WpCedOFk.dkqSpvm","createdAt":"2023-10-21T12:41:46.989Z","updatedAt":"2023-10-21T12:41:47.100Z"}