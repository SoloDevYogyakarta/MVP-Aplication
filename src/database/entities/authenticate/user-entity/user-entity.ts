import { nanoid } from 'nanoid';
import { DataTypes, Model, ModelAttributes, Optional } from 'sequelize';
import { SuperTest } from 'supertest';
import { BaseEntity } from '../../base-entity/base-entity';
import { sequelize } from '../../entity';

export class UserEntity extends BaseEntity {
  plat_number: string;
  phone_number: string;
  name: string;
  motor: string;
  year_production: number;
  address: string;
  role: string;
  password: string;
}

export type UserCreationAttribute = Optional<UserEntity, 'id'>;
export interface UserInstance
  extends Model<UserCreationAttribute, UserEntity>,
    UserEntity {}
export const userEntity = sequelize.define<UserInstance>(
  'USER',
  {
    plat_number: DataTypes.STRING,
    phone_number: DataTypes.STRING,
    name: DataTypes.STRING,
    motor: DataTypes.STRING,
    year_production: DataTypes.INTEGER,
    address: DataTypes.STRING,
    role: DataTypes.STRING,
    password: DataTypes.STRING,
  } as ModelAttributes<UserInstance>,
  {
    tableName: 'USER',
    schema: 'AUTHENTICATE',
  },
);
