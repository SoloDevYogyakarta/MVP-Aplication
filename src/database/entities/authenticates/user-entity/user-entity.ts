import { DataTypes, Model, ModelAttributes, Optional } from 'sequelize';
import { nanoid } from 'nanoid';
import bcrypt from 'bcrypt';
import { sequelize } from '../../entity';
import { BaseEntity } from '../../base-entity/base-entity';

export class UserEntity extends BaseEntity {
  username: string;
  email: string;
  phone_number: number;
  password: string;
  role: string;
  file_id: string;
}

export type UserCreationAttribute = Optional<UserEntity, 'id'>;
export interface UserInstance
  extends Model<UserCreationAttribute, UserEntity>,
    UserEntity {}
export const userEntity = sequelize.define<UserInstance>(
  'USER',
  {
    public_id: DataTypes.STRING,
    username: DataTypes.STRING,
    email: DataTypes.STRING,
    phone_number: DataTypes.INTEGER,
    role: DataTypes.STRING,
    password: DataTypes.STRING,
    file_id: DataTypes.STRING,
  } as ModelAttributes<UserInstance>,
  {
    tableName: 'USER',
    schema: 'AUTHENTICATE',
    hooks: {
      beforeCreate(attributes, options) {
        const instance = attributes;
        instance.public_id = nanoid();
        instance.password = bcrypt.hashSync(instance.password, 10);
      },
    },
  },
);
