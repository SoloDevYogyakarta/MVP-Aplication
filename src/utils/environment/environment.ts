import { config } from 'dotenv';

config();

export const environment: NodeJS.ProcessEnv = process.env;
