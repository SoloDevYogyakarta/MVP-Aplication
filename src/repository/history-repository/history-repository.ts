import { Injectable, Logger } from '@nestjs/common';
import { sequelize } from '../../database/entities/entity';
import {
  userAssociate,
  userAttribute,
  userHistoryInclude,
} from '../../database/associates/user-associate/user-associate';
import { UserInstance } from '../../database/entities/authenticate/user-entity/user-entity';
import { OrderEntity } from '../../database/entities/services/order-entity/order-entity';

@Injectable()
export class HistoryRepository {
  private readonly logger = new Logger(HistoryRepository.name);

  async findAll(): Promise<UserInstance[]> {
    this.logger.log(HistoryRepository.name);
    let result = await userAssociate.findAll({
      attributes: userAttribute,
      include: userHistoryInclude,
    });
    for (const values of result) {
      const visit = await this.visit(values.id);
      const total = await this.total(values.id);
      const totalOrder = await this.totalOrder(values.id);
      result = JSON.parse(JSON.stringify(result)).map((item: UserInstance) => {
        if (item.id === values.id) {
          item['visit'] = Number(visit);
          item['total'] = total;
        }
        if (totalOrder.length) {
          item['order'] = (
            item as unknown as { order: OrderEntity[] }
          ).order.map((ord) => {
            const total_order = totalOrder.find((i) => i.id === ord.id);
            if (total_order) {
              if (item.id === ord.user_id) {
                ord['total'] = total_order?.total
                  ? Number(total_order.total)
                  : 0;
              }
            }
            return ord;
          });
        }
        return item;
      });
    }
    return result;
  }

  async findOne(id: number): Promise<UserInstance> {
    this.logger.log(HistoryRepository.name);
    let result = await userAssociate.findOne({
      where: { id },
      attributes: userAttribute,
      include: userHistoryInclude,
    });
    result = JSON.parse(JSON.stringify(result));
    for (const values of (result as unknown as { order: OrderEntity[] })
      .order) {
      const total = await this.totalOrder(values.id);
      result['order'] = (
        result as unknown as { order: OrderEntity[] }
      ).order.map((item) => {
        if (item.id === values.id) {
          item['total'] = total;
        }
        return item;
      });
    }
    result['visit'] = Number(await this.visit(id));
    result['total'] = await this.total(id);
    return result;
  }

  async visit(id: number): Promise<number> {
    const query = await sequelize.query(
      `SELECT a.user_id, COUNT(a.user_id) as visit FROM "PRODUCTS"."ORDER" AS a WHERE a.user_id = '${id}' GROUP BY a.user_id `,
    );
    const visit = query[0].find(
      (item: { user_id: number; visit: number }) =>
        Number(item.user_id) === Number(id),
    ) as { user_id: number; visit: number };
    return visit?.visit ? visit.visit : 0;
  }

  async totalOrder(
    id: number,
  ): Promise<{ id: number; user_id: number; total: string }[]> {
    const query = await sequelize.query(`
    SELECT a.id,a.user_id,SUM(CAST(b.price as int)) as total FROM "PRODUCTS"."ORDER" AS a
    LEFT JOIN "SERVICES"."HISTORY" AS b ON b.order_id = a.id WHERE a.user_id = ${id}
    GROUP BY a.id;
    `);
    return query[0] as unknown as {
      id: number;
      user_id: number;
      total: string;
    }[];
  }

  async total(id: number): Promise<number> {
    const query = await sequelize.query(`
      SELECT a.user_id,SUM(CAST(b.price as int)) as total FROM "PRODUCTS"."ORDER" AS a
      LEFT JOIN "SERVICES"."HISTORY" AS b ON b.order_id = a.id WHERE a.user_id = ${id}
      GROUP BY a.id
    `);
    const result = query[0]
      .map((item: { total: string }) => (item.total === null ? 0 : item.total))
      ?.reduce((arr: unknown, curr: unknown) => {
        return Number(arr) + Number(curr);
      }, 0);
    return result ? Number(result) : 0;
  }
}
