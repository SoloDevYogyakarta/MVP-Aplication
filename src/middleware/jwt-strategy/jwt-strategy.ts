import { Injectable } from '@nestjs/common';
import { PassportStrategy } from '@nestjs/passport';
import { ExtractJwt, Strategy, StrategyOptions } from 'passport-jwt';
import { environment } from '../../utils/environment/environment';

@Injectable()
export class JwtStrategy extends PassportStrategy(Strategy) {
  constructor() {
    super({
      jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
      ignoreExpiration: false,
      secretOrKey: environment['SECRET'],
    } as Partial<StrategyOptions>);
  }
}
