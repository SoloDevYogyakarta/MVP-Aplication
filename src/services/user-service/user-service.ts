import { HttpException, HttpStatus, Injectable, Logger } from '@nestjs/common';
import bcrypt from 'bcrypt';
import { omit } from 'lodash';
import { userEntity } from '../../database/entities/authenticate/user-entity/user-entity';
import {
  ChangePasswordField,
  LoginField,
  RegisterField,
} from '../../dto/user-dto/user-dto';
import { environment } from '../../utils/environment/environment';
import { JwtService } from '@nestjs/jwt';
import { userHistoryEntity } from '../../database/entities/authenticate/history-entity/history-entity';

@Injectable()
export class UserService {
  private readonly logger = new Logger(UserService.name);

  constructor(private readonly jwtService: JwtService) {}

  async login(body: LoginField) {
    this.logger.log(UserService.name);
    const findOne = await userEntity.findOne({
      where: { plat_number: body.token },
    });
    if (!findOne) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: 'Invalid username or password',
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    const check = await bcrypt.compareSync(body.password, findOne.password);
    if (!check) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: 'Invalid username or password',
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    const access_token = this.jwtService.sign(
      {
        data: JSON.parse(JSON.stringify(findOne)),
      },
      { secret: environment['SECRET'] },
    );
    return { status: HttpStatus.OK, access_token };
  }

  async create(body: RegisterField) {
    this.logger.log(UserService.name);
    const findOne = await userEntity.findOne({
      where: { plat_number: body.plat_number },
    });
    if (findOne) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: 'Plat number already exists',
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    if (body.password !== body.confirmation) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: "Password don't match, please check again",
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    const create = await userEntity.create(
      omit({ ...body, password: await bcrypt.hashSync(body.password, 15) }, [
        'confirmation',
      ]),
    );
    create.save();
    return {
      status: HttpStatus.CREATED,
      message: 'Account has been create',
      result: create,
    };
  }

  async update(id: number, body: RegisterField) {
    const findOne = await userEntity.findOne({ where: { id } });
    if (!findOne) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: 'Account not found',
        },
        HttpStatus.BAD_REQUEST,
      );
    }

    const history = await userHistoryEntity.create({
      ...omit(JSON.parse(JSON.stringify(findOne)), [
        'id',
        'public_id',
        'role',
        'password',
        'createdAt',
        'updatedAt',
      ]),
      user_id: findOne.id,
    });
    history.save();
    findOne.update(
      omit(body, ['motor', 'year_production', 'password', 'confirmation']),
      { where: { id } },
    );
    return { status: HttpStatus.OK, message: 'Account has been update' };
  }

  async changePassword(id: number, body: ChangePasswordField) {
    const findOne = await userEntity.findOne({ where: { id } });
    if (!findOne) {
      throw new HttpException(
        {
          status: HttpStatus.INTERNAL_SERVER_ERROR,
          message: 'false',
        },
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
    const check = await bcrypt.compareSync(body.old_password, findOne.password);
    if (!check) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: 'Wrong password',
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    if (body.password !== body.confirmation) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: "Password don't match, please check again",
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    findOne.password = await bcrypt.hashSync(body.password, 15);
    findOne.save();
    return { status: HttpStatus.OK, message: 'Password has been update' };
  }

  async destroy(id: number) {
    const findOne = await userEntity.findOne({ where: { id } });
    if (!findOne) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: 'Account not found',
        },
        HttpStatus.BAD_REQUEST,
      );
    }

    const history = await userHistoryEntity.findAll({
      where: { user_id: findOne.id },
    });

    for (const values of history) {
      values?.destroy();
    }
    findOne.destroy();
    return { status: HttpStatus.OK, message: 'Account has been delete' };
  }
}
