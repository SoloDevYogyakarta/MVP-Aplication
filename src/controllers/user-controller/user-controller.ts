import {
  Body,
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Logger,
  Post,
  Req,
  Res,
  UseGuards,
} from '@nestjs/common';
import { omit } from 'lodash';
import { Response } from 'express';
import {
  ChangePasswordField,
  LoginField,
  RegisterField,
} from '../../dto/user-dto/user-dto';
import { UserRepository } from '../../repository/user-repository/user-repository';
import { UserService } from '../../services/user-service/user-service';
import { AuthGuard } from '../../middleware/auth-guard/auth-guard';
import { CustomRequest } from '../../types/custom-request.type';

@Controller('user')
export class UserController {
  private readonly logger = new Logger(UserController.name);

  constructor(
    private readonly repository: UserRepository,
    private readonly service: UserService,
  ) {}

  @Post('login')
  @HttpCode(HttpStatus.OK)
  async login(@Body() body: LoginField, @Res() res: Response) {
    this.logger.log(UserController.name);
    const result = await this.service.login(body);
    return res.status(result.status).json(result);
  }

  @Post()
  @HttpCode(HttpStatus.CREATED)
  async create(@Body() body: RegisterField, @Res() res: Response) {
    this.logger.log(UserController.name);
    const result = await this.service.create(body);
    return res.status(result.status).json(omit(result, ['result']));
  }

  @UseGuards(AuthGuard)
  @Post('update/:id')
  @HttpCode(HttpStatus.OK)
  async update(
    @Req() req: CustomRequest,
    @Body() body: RegisterField,
    @Res() res: Response,
  ) {
    const result = await this.service.update(req.params.id, body);
    return res.status(result.status).json(result);
  }

  @UseGuards(AuthGuard)
  @Post('update/password/:id')
  @HttpCode(HttpStatus.OK)
  async updatePassword(
    @Req() req: CustomRequest,
    @Body() body: ChangePasswordField,
    @Res() res: Response,
  ) {
    const result = await this.service.changePassword(req.params.id, body);
    return res.status(result.status).json(result);
  }

  @UseGuards(AuthGuard)
  @Get()
  @HttpCode(HttpStatus.OK)
  async all(@Res() res: Response) {
    const result = await this.repository.findAll();
    return res.status(HttpStatus.OK).json(result);
  }

  @UseGuards(AuthGuard)
  @Get(':id')
  @HttpCode(HttpStatus.OK)
  async detail(@Req() req: CustomRequest, @Res() res: Response) {
    const result = await this.repository.findOne(req.params.id);
    return res.status(HttpStatus.OK).json(result);
  }

  @UseGuards(AuthGuard)
  @Get('current/user')
  @HttpCode(HttpStatus.OK)
  async me(@Req() req: CustomRequest, @Res() res: Response) {
    return res.status(HttpStatus.OK).json(req.user);
  }
}
