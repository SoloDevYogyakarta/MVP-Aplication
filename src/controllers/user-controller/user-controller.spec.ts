import { HttpStatus, INestApplication } from '@nestjs/common';
import { JwtModule } from '@nestjs/jwt';
import { Test } from '@nestjs/testing';
import { environment } from '../../utils/environment/environment';
import { UserService } from '../../services/user-service/user-service';
import { UserController } from './user-controller';
import supertest from 'supertest';
import {
  ChangePasswordField,
  LoginField,
  RegisterField,
} from 'src/dto/user-dto/user-dto';
import { faker } from '@faker-js/faker';
import { UserInstance } from '../../database/entities/authenticate/user-entity/user-entity';
import { createpath } from '../../utils/system/system';
import { getfield } from '../../utils/get-field/get-field';
import { JwtStrategy } from '../../middleware/jwt-strategy/jwt-strategy';
import { HistoryRepository } from '../../repository/history-repository/history-repository';

describe('UserController', () => {
  let app: INestApplication;
  let token!: string;
  let user!: UserInstance;
  let destroy_id!: number;
  let single_destroy_id!: number;
  let admin!: UserInstance;
  let visit!: number;

  beforeEach(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [
        JwtModule.register({
          secret: environment['SECRET'],
          signOptions: {
            expiresIn: '7d',
          },
        }),
      ],
      controllers: [UserController],
      providers: [UserService, HistoryRepository, JwtStrategy],
    }).compile();

    app = moduleRef.createNestApplication();
    await app.init();
  });

  it('should to be defined', () => expect(app).toBeDefined());

  it('render correctly', () => expect(app.getHttpServer()).toMatchSnapshot());

  try {
    token = getfield('token');
  } catch (err) {
    // empty
  }
  try {
    user = getfield('user-http-entity');
  } catch (err) {
    // empty
  }
  try {
    admin = getfield('user-admin-entity');
  } catch (err) {
    // empty
  }
  try {
    single_destroy_id = getfield('user-entity').id;
  } catch (err) {
    // empty
  }
  try {
    destroy_id = getfield('user-history-destroy-entity').user_id;
  } catch (err) {
    // empty
  }
  try {
    visit = getfield('visit');
  } catch (err) {
    // empty
  }

  it('http::user create', async () => {
    await supertest(app.getHttpServer())
      .post('/users')
      .set('content-type', 'application/json')
      .send({
        plat_number: `BG ${Math.floor(1000 + Math.random() * 9999)} JAK`,
        phone_number: faker.phone.number(),
        name: faker.person.fullName(),
        motor: faker.commerce.productName(),
        year_production: Math.floor(1000 + Math.random() * 2023),
        address: faker.location.streetAddress(),
        password: 'password',
        confirmation: 'password',
      } as Partial<RegisterField>)
      .expect(HttpStatus.CREATED)
      .then((res) =>
        expect(res.body).toEqual({
          status: HttpStatus.CREATED,
          message: 'Account has been create',
        }),
      );
  });

  if (user?.id) {
    it('http::user create already exists', async () => {
      await supertest(app.getHttpServer())
        .post('/users')
        .set('content-type', 'application/json')
        .send({
          plat_number: user.plat_number,
          phone_number: faker.phone.number(),
          name: faker.person.fullName(),
          motor: faker.commerce.productName(),
          year_production: Math.floor(1000 + Math.random() * 2023),
          address: faker.location.streetAddress(),
          password: 'password',
          confirmation: 'password',
        } as Partial<RegisterField>)
        .expect(HttpStatus.BAD_REQUEST)
        .then((res) =>
          expect(res.body).toEqual({
            status: HttpStatus.BAD_REQUEST,
            message: 'Plat number already exists',
          }),
        );
    });
  }

  it('http::user create already exists', async () => {
    await supertest(app.getHttpServer())
      .post('/users')
      .set('content-type', 'application/json')
      .send({
        plat_number: admin.plat_number,
        phone_number: faker.phone.number(),
        name: faker.person.fullName(),
        motor: faker.commerce.productName(),
        year_production: Math.floor(1000 + Math.random() * 2023),
        address: faker.location.streetAddress(),
        password: 'password',
        confirmation: 'password',
      } as Partial<RegisterField>)
      .expect(HttpStatus.BAD_REQUEST)
      .then((res) =>
        expect(res.body).toEqual({
          status: HttpStatus.BAD_REQUEST,
          message: 'Plat number already exists',
        }),
      );
  });

  it("http::user create password don't match ", async () => {
    await supertest(app.getHttpServer())
      .post('/users')
      .set('content-type', 'application/json')
      .send({
        plat_number: `BG ${Math.floor(1000 + Math.random() * 9999)} JAK`,
        phone_number: faker.phone.number(),
        name: faker.person.fullName(),
        motor: faker.commerce.productName(),
        year_production: Math.floor(1000 + Math.random() * 2023),
        address: faker.location.streetAddress(),
        password: 'password',
        confirmation: 'passwordqdqwd',
      } as Partial<RegisterField>)
      .expect(HttpStatus.BAD_REQUEST)
      .then((res) =>
        expect(res.body).toEqual({
          status: HttpStatus.BAD_REQUEST,
          message: "Password don't match, please check again",
        }),
      );
  });

  if (admin?.id) {
    it('http::user login', async () => {
      await supertest(app.getHttpServer())
        .post('/users/login')
        .set('content-type', 'application/json')
        .send({
          token: admin.plat_number,
          password: 'password',
        } as Partial<LoginField>)
        .expect(HttpStatus.OK)
        .then((res) => {
          createpath('../folder-text/token.txt', res.body.access_token);
          expect(res.body.access_token).not.toEqual(null);
        });
    });
  }

  if (user?.id) {
    it('http::user login role member', async () => {
      await supertest(app.getHttpServer())
        .post('/users/login')
        .set('content-type', 'application/json')
        .send({
          token: user.plat_number,
          password: 'password',
        } as Partial<LoginField>)
        .expect(HttpStatus.OK)
        .then((res) => {
          createpath('../folder-text/token_member.txt', res.body.access_token);
          expect(res.body.access_token).not.toEqual(null);
        });
    });

    it('http::user invalid password login', async () => {
      await supertest(app.getHttpServer())
        .post('/users/login')
        .set('content-type', 'application/json')
        .send({
          token: user.plat_number,
          password: 'passwodqwdqrd',
        } as Partial<LoginField>)
        .expect(HttpStatus.BAD_REQUEST)
        .then((res) =>
          expect(res.body).toEqual({
            status: HttpStatus.BAD_REQUEST,
            message: 'Invalid username or password',
          }),
        );
    });
  }

  it('http::user invalid login', async () => {
    await supertest(app.getHttpServer())
      .post('/users/login')
      .set('content-type', 'application/json')
      .send({
        token: 'dqdqw',
        password: 'password',
      } as Partial<LoginField>)
      .expect(HttpStatus.BAD_REQUEST)
      .then((res) =>
        expect(res.body).toEqual({
          status: HttpStatus.BAD_REQUEST,
          message: 'Invalid username or password',
        }),
      );
  });

  if (token && user?.id) {
    it('http::user update', async () => {
      await supertest(app.getHttpServer())
        .post(`/users/update/${user.id}`)
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .send({
          plat_number: user.plat_number,
          phone_number: faker.phone.number(),
          name: faker.person.fullName(),
          address: faker.location.streetAddress(),
        } as Partial<RegisterField>)
        .expect(HttpStatus.OK)
        .then((res) =>
          expect(res.body).toEqual({
            status: HttpStatus.OK,
            message: 'Account has been update',
          }),
        );
    });

    it('http::user invalid update', async () => {
      await supertest(app.getHttpServer())
        .post(`/users/update/0000`)
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .send({
          plat_number: user.plat_number,
          phone_number: faker.phone.number(),
          name: faker.person.fullName(),
          address: faker.location.streetAddress(),
        } as Partial<RegisterField>)
        .expect(HttpStatus.BAD_REQUEST)
        .then((res) =>
          expect(res.body).toEqual({
            status: HttpStatus.BAD_REQUEST,
            message: 'Account not found',
          }),
        );
    });

    it('http::user update password', async () => {
      await supertest(app.getHttpServer())
        .post(`/users/update/password/${user.id}`)
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .send({
          plat_number: user.plat_number,
          old_password: 'password',
          password: 'password',
          confirmation: 'password',
        } as Partial<ChangePasswordField>)
        .expect(HttpStatus.OK)
        .then((res) =>
          expect(res.body).toEqual({
            status: HttpStatus.OK,
            message: 'Password has been update',
          }),
        );
    });

    it('http::user invalid update password', async () => {
      await supertest(app.getHttpServer())
        .post(`/users/update/password/00`)
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .send({
          plat_number: user.plat_number,
          old_password: 'password',
          password: 'password',
          confirmation: 'password',
        } as Partial<ChangePasswordField>)
        .expect(HttpStatus.INTERNAL_SERVER_ERROR)
        .then((res) =>
          expect(res.body).toEqual({
            status: HttpStatus.INTERNAL_SERVER_ERROR,
            message: 'false',
          }),
        );
    });

    it('http::user update wrong password', async () => {
      await supertest(app.getHttpServer())
        .post(`/users/update/password/${user.id}`)
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .send({
          plat_number: user.plat_number,
          old_password: 'passwdqdqord',
          password: 'password',
          confirmation: 'password',
        } as Partial<ChangePasswordField>)
        .expect(HttpStatus.BAD_REQUEST)
        .then((res) =>
          expect(res.body).toEqual({
            status: HttpStatus.BAD_REQUEST,
            message: 'Wrong password',
          }),
        );
    });

    it('http::user update password dont match', async () => {
      await supertest(app.getHttpServer())
        .post(`/users/update/password/${user.id}`)
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .send({
          plat_number: user.plat_number,
          old_password: 'password',
          password: 'password',
          confirmation: 'pasdqwdqsword',
        } as Partial<ChangePasswordField>)
        .expect(HttpStatus.BAD_REQUEST)
        .then((res) =>
          expect(res.body).toEqual({
            status: HttpStatus.BAD_REQUEST,
            message: "Password don't match, please check again",
          }),
        );
    });

    it('http::history all', async () => {
      await supertest(app.getHttpServer())
        .get('/users')
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .expect(HttpStatus.OK)
        .then((res) => {
          expect(res.body.length).not.toEqual(0);
        });
    });

    it('http::history detail', async () => {
      await supertest(app.getHttpServer())
        .get(`/users/${user.id}/services`)
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .then((res) => {
          createpath('../folder-text/history-http-entity.txt', res.body);
          expect(res.body.id).toEqual(user.id);
        });
    });

    if (visit) {
      it('http::history detail visit', async () => {
        await supertest(app.getHttpServer())
          .get(`/users/${visit}/services`)
          .set('content-type', 'application/json')
          .set('Authorization', `Bearer ${token}`)
          .then(async (res) => {
            createpath('../folder-text/history-http-entity.txt', res.body);
            expect(res.body.visit).not.toEqual(0);
          });
      });
    }

    it('http::user detail unauthorization', async () => {
      await supertest(app.getHttpServer())
        .get(`/users/${user.id}/services`)
        .set('content-type', 'application/json')
        .expect(HttpStatus.UNAUTHORIZED)
        .then((res) =>
          expect(res.body).toEqual({
            statusCode: HttpStatus.UNAUTHORIZED,
            message: 'Unauthorized',
          }),
        );
    });

    it('http::user me', async () => {
      await supertest(app.getHttpServer())
        .get('/users/current/user')
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .expect(HttpStatus.OK)
        .then((res) => expect(res.body).not.toEqual(null));
    });

    it('http::user destroy', async () => {
      await supertest(app.getHttpServer())
        .delete(`/users/${single_destroy_id}`)
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .expect(HttpStatus.OK)
        .then((res) =>
          expect(res.body).toEqual({
            status: HttpStatus.OK,
            message: 'Account has been delete',
          }),
        );
    });

    it('http::user destroy with relationship', async () => {
      await supertest(app.getHttpServer())
        .delete(`/users/${destroy_id}`)
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .expect(HttpStatus.OK)
        .then((res) =>
          expect(res.body).toEqual({
            status: HttpStatus.OK,
            message: 'Account has been delete',
          }),
        );
    });

    it('http::user invalid destroy', async () => {
      await supertest(app.getHttpServer())
        .delete(`/users/0000`)
        .set('content-type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .expect(HttpStatus.BAD_REQUEST)
        .then((res) =>
          expect(res.body).toEqual({
            status: HttpStatus.BAD_REQUEST,
            message: 'Account not found',
          }),
        );
    });
  }
});
