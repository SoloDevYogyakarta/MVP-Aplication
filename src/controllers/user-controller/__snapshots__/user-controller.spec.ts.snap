// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserController render correctly 1`] = `
Server {
  "_connections": 0,
  "_events": {
    "connection": [Function],
    "request": [Function],
  },
  "_eventsCount": 2,
  "_handle": null,
  "_maxListeners": undefined,
  "_unref": false,
  "_usingWorkers": false,
  "_workers": [],
  "allowHalfOpen": true,
  "connectionsCheckingInterval": 30000,
  "headersTimeout": 60000,
  "httpAllowHalfOpen": false,
  "insecureHTTPParser": undefined,
  "keepAlive": false,
  "keepAliveInitialDelay": 0,
  "keepAliveTimeout": 5000,
  "maxHeaderSize": undefined,
  "maxHeadersCount": null,
  "maxRequestsPerSocket": 0,
  "noDelay": true,
  "pauseOnConnect": false,
  "requestTimeout": 300000,
  "timeout": 0,
  Symbol(IncomingMessage): [Function],
  Symbol(ServerResponse): [Function],
  Symbol(kCapture): false,
  Symbol(async_id_symbol): -1,
  Symbol(http.server.connections): ConnectionsList {},
  Symbol(http.server.connectionsCheckingInterval): Timeout {
    "_destroyed": false,
    "_idleNext": Timeout {
      "_destroyed": false,
      "_idleNext": TimersList {
        "_idleNext": [Circular],
        "_idlePrev": [Circular],
        "expiry": 34608,
        "id": -9007199254740988,
        "msecs": 30000,
        "priorityQueuePosition": 2,
      },
      "_idlePrev": [Circular],
      "_idleStart": 4608,
      "_idleTimeout": 30000,
      "_onTimeout": [Function],
      "_repeat": 30000,
      "_timerArgs": undefined,
      Symbol(refed): false,
      Symbol(kHasPrimitive): false,
      Symbol(asyncId): 19,
      Symbol(triggerId): 0,
    },
    "_idlePrev": TimersList {
      "_idleNext": [Circular],
      "_idlePrev": Timeout {
        "_destroyed": false,
        "_idleNext": [Circular],
        "_idlePrev": [Circular],
        "_idleStart": 4608,
        "_idleTimeout": 30000,
        "_onTimeout": [Function],
        "_repeat": 30000,
        "_timerArgs": undefined,
        Symbol(refed): false,
        Symbol(kHasPrimitive): false,
        Symbol(asyncId): 19,
        Symbol(triggerId): 0,
      },
      "expiry": 34608,
      "id": -9007199254740988,
      "msecs": 30000,
      "priorityQueuePosition": 2,
    },
    "_idleStart": 4631,
    "_idleTimeout": 30000,
    "_onTimeout": [Function],
    "_repeat": 30000,
    "_timerArgs": undefined,
    Symbol(refed): false,
    Symbol(kHasPrimitive): false,
    Symbol(asyncId): 23,
    Symbol(triggerId): 0,
  },
  Symbol(kUniqueHeaders): null,
}
`;
